### ISSUE TITLE
app crashes on getAccounts()

### ISSUE THREAD

**Issue by jbloomfield2** [AzureAD/microsoft-authentication-library-for-android#495]
I am following the sample app to try and get silent login, but my app crashes on the getAccounts() function. if i clear the app storage no crash but if there is a previous login it crashes with the following stack trace

```
* Attempt to get a user and acquireTokenSilent
         * If this fails we do an interactive request
         */
        List<IAccount> accounts = null;
try {
            accounts = myApp.getAccounts(); //crashes here

            if (accounts != null && accounts.size() == 1) {
                /* We have 1 account */
               ...
}...
```
```java
java.lang.RuntimeException: Unable to resume activity {jbloomfield.macroapp/jbloomfield.macroapp.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Object.toString()' on a null object reference
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3822)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3854)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:51)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1816)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:495)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:859)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Object.toString()' on a null object reference
        at com.microsoft.identity.common.internal.cache.AccountCredentialCache.getAccountsWithKeys(AccountCredentialCache.java:168)
        at com.microsoft.identity.common.internal.cache.AccountCredentialCache.getAccounts(AccountCredentialCache.java:184)
        at com.microsoft.identity.common.internal.cache.AccountCredentialCache.getAccountsFilteredBy(AccountCredentialCache.java:205)
        at com.microsoft.identity.common.internal.cache.MsalOAuth2TokenCache.getAccounts(MsalOAuth2TokenCache.java:337)
        at com.microsoft.identity.client.PublicClientApplication.getAccounts(PublicClientApplication.java:330)
        at jbloomfield.macroapp.AuthManager.getAccs(AuthManager.java:124)
        at jbloomfield.macroapp.MainActivity.onResume(MainActivity.java:129)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1413)
        at android.app.Activity.performResume(Activity.java:7301)
```

**Comment by heidijinxujia** [comment]
Not sure which branch are you using.

The API `PublicClientApplication.getAccounts()` has been replaced by the async API `PublicClientApplication.getAccounts(final AccountsLoadedListener callback)`. You need to implement the interface like the following

```java
myApp.getAccounts(new PublicClientApplication.AccountsLoadedListener() {
	@Override
	//Called once Accounts have been loaded from the cache.
	public void onAccountsLoaded((List<IAccount> accounts) {
		//Your code here...
	}
});
```

The code is currently in dev branch and will be publish in the next MSAL release.

**Comment by 4gus71n** [comment]
@heidijinxujia I'm having the exact same issue, how can I include the dependency to the dev branch in my Gradle file?

Something like:
```
dependencies {
	        implementation 'com.github.AzureAD:microsoft-authentication-library-for-android:dev'
	}
```

There's no other way to workaround this bug other than using the `dev` branch? I don't feel comfortable using a development branch for the library.

**Comment by Larrsenn** [comment]
@heidijinxujia when is the next MSAL release then? Because this is kind of a blocker for some, so would be super sweet to get this out asap.

**Comment by iambmelt** [comment]
I don't believe this issue will be resolved simply by switching to the new API, as it's built on the same fundamentals of the old `getAccounts` API

**Comment by jbloomfield2** [comment]
Some additional details: the code works fine in the emulator but crashes on my device. Both are Android 9.  By stepping into the 
com.microsoft.identity.common.internal.cache.AccountCredentialCache.getAccountsWithKeys(AccountCredentialCache.java:168)

 function there is a loop which retrieves key/value pairs. The value are null on my device and look fine in the emulator. This looks like the cached credentials are not being saved correctly

**Comment by Larrsenn** [comment]
Same error for me jbloomfield, haven't been able to test using my device for a week now.

**Comment by iambmelt** [comment]
Thanks for uploading the logs, @jbloomfield2 

From the logs:
```
E/SharedPreferencesFileManager:encryptDecryptInternal:  [2019-02-05 08:31:31 - {"thread_id":"2","correlation_id":"8e3484b6-5ca4-4c05-905b-aefc82df74a3"}] Failed to decrypt value
```

I can see [this block of code is throwing an exception during decryption](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesFileManager.java#L202-L220).

Anything 'special' you can share about this device/emulator? Any particular repro steps?

To my knowledge, there have been Android P issues with `KeyStore` before, some of these were in Preview:
https://issuetracker.google.com/issues/78206414
https://issuetracker.google.com/issues/74355119

When saving or restoring the accounts, are you seeing error messages in the log? Anything which might indicate that it is failing to encrypt or load the decryption keys?

**Comment by DanyGZ** [comment]
I have the same issue. The api seems to work fine in some devices. But is crashing in a few devices. After I clean cache or application data, the api works fine. Is there any update or any way to fix this?

**Comment by iambmelt** [comment]
@DanyGZ A fix for this will debut in an upcoming release but I don't have an ETA at this time, unfortunately. This has been one of those issues where fixing some devices seems to break others... any additional information you can provide about where you're seeing issues would be great... OS version numbers, device make/model, locale, etc.

**Comment by DanyGZ** [comment]
@iambmelt  I am running on a galaxy s6 (android 6.0.1). This is the only device in which I'm having the crash.

**Comment by jjerome00** [comment]
Is there a way to clear the library's cache?  There does not seem to be a way to do it.  Is the only option to clear the entire app's data?  i.e. `activity.activityManager.clearApplicationUserData()`  ?

**Comment by iambmelt** [comment]
@jjerome00 Our expected usage pattern is for apps to call `PublicClientApplication#getAccounts()` and then call `PublicClientApplication#removeAccount()` on each of those. This will clear tokens stored on disk.

SSO cookies in the WebView or browser will be preserved. To clear SSO cookies in the browser is a slightly different story and depends on whether or not your app is using the embedded WebView or delegating to a browser to auth.

For embedded WebView, clearing the application data through the settings (or Activity Manager) will do the trick. For browser-based auth, the cache for that specific browser must be cleared which has the unfortunate side-effect of taking _all other_ cookies with it too - likely not the experience you would be trying to achieve.

Feel free to reach out with further questions or if this doesn't hit the necessary scenario

**Comment by haphu4247** [comment]
i have the same issue like this, it happened on window, but it work on Mac, i feel confused about how to explan the reason it work!! hope this help!!

**Comment by jjerome00** [comment]
@iambmelt thanks for your through reply.  In the vein of this issue, calling `getAccounts()` causes the crash.  I was looking for a way to clear the cached accounts without querying for them.  

This way if I'm in a bad state and `getAccounts()` fails, I could prevent an app crash by clearing my broken cache, and just ask the user to log in again.  At this point my last resort is to clear user data via Android, which will shut down the app.

If it's not possible, that's fine.

**Comment by 4gus71n** [comment]
Hey guys, I've implemented a workaround for this issue, I'm just catching the exception and when that happens I:

```
...
        } catch (e: Exception) {
            val msalPrefs = context.getSharedPreferences("com.microsoft.identity.client.account_credential_cache", Context.MODE_PRIVATE)
            msalPrefs.edit().clear().commit()
            Log.w("MSAL", e)
            callback.onError()
        }
...
```

I know that this is a super hacky solution, but It's the only workaround that I've been able to find so far. If any of you can confirm that this fixes the issue for you as well, that would be super useful for me.

Thanks!

**Comment by DanyGZ** [comment]
@4gus71n It seems to work fine, at my first attempt the application didn't crash and opened the login correctly.

**Comment by iambmelt** [comment]
Related https://github.com/AzureAD/microsoft-authentication-library-common-for-android/issues/40

**Comment by joshfriend** [comment]
I've been seeing this crash a _lot_ with v0.2.2

**Comment by heidijinxujia** [comment]
@joshfriend are you still encountering this issue with the latest version?

**Comment by predalpha** [comment]
The issue is still present on msal 0.2.2

**Comment by iambmelt** [comment]
This issue should be fixed with `1.0.0` -- please open a new issue if you have any problems
